File: pem_to_csv.py
#!/usr/bin/env python3
"""
PEM to CSV Converter
Extracts XML content from PEM files using ASN.1 parsing and converts to CSV.
"""

import os
import sys
import argparse
import base64
import re
import csv
from typing import Dict, List, Any, Optional
import xml.etree.ElementTree as ET
from pathlib import Path

try:
    from asn1crypto import cms, core
except ImportError:
    print("❌ asn1crypto library not found. Installing...")
    os.system("pip install asn1crypto")
    from asn1crypto import cms, core


def find_pem_files(directory: str = ".") -> List[str]:
    """Find all PEM files in the specified directory."""
    pem_files = []
    for file in os.listdir(directory):
        if file.endswith(".pem"):
            pem_files.append(file)
    return sorted(pem_files)


def select_pem_file(file_path: Optional[str] = None) -> str:
    """Interactive file selection or use provided file path."""
    if file_path:
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")
        return file_path
    
    print("🔍 Looking for PEM files in current directory...")
    pem_files = find_pem_files()
    
    if not pem_files:
        print("❌ No PEM files found in current directory")
        print("Usage: python pem_to_csv.py --file <path-to-pem-file>")
        sys.exit(1)
    
    print("\n📁 Available PEM files:")
    for i, file in enumerate(pem_files, 1):
        print(f"  {i}. {file}")
    
    while True:
        try:
            choice = input("\n📝 Enter the number of the file you want to process (or press Enter for file 1): ").strip()
            if not choice:
                choice = "1"
            
            selection = int(choice) - 1
            if 0 <= selection < len(pem_files):
                return pem_files[selection]
            else:
                print("❌ Invalid selection. Please try again.")
        except ValueError:
            print("❌ Please enter a valid number.")


def extract_xml_from_pem(pem_content: str) -> str:
    """Extract XML content from PEM file using ASN.1 parsing."""
    try:
        # Remove PEM headers and decode base64
        lines = pem_content.strip().split('\n')
        base64_content = ""
        inside_pem = False
        
        for line in lines:
            line = line.strip()
            if line.startswith("-----BEGIN"):
                inside_pem = True
                continue
            elif line.startswith("-----END"):
                inside_pem = False
                break
            elif inside_pem:
                base64_content += line
        
        if not base64_content:
            raise ValueError("No base64 content found in PEM file")
        
        # Decode base64 to DER
        der_data = base64.b64decode(base64_content)
        
        # Try to parse as CMS/PKCS#7
        try:
            cms_obj = cms.ContentInfo.load(der_data)
            
            # Extract content
            if cms_obj['content_type'].dotted == '1.2.840.113549.1.7.1':  # data
                content = cms_obj['content'].contents
                content_str = content.decode('utf-8', errors='ignore')
            elif cms_obj['content_type'].dotted == '1.2.840.113549.1.7.2':  # signedData
                signed_data = cms_obj['content']
                if 'encap_content_info' in signed_data and signed_data['encap_content_info']['content']:
                    content = signed_data['encap_content_info']['content'].contents
                    content_str = content.decode('utf-8', errors='ignore')
                else:
                    content_str = str(signed_data)
            else:
                content_str = str(cms_obj)
                
        except Exception as e:
            print(f"⚠️  CMS parsing failed: {e}")
            # Fallback: try to find XML directly in the DER data
            content_str = der_data.decode('utf-8', errors='ignore')
        
        # Search for XML content
        xml_pattern = r'<\?xml.*?</[^>]+>'
        xml_match = re.search(xml_pattern, content_str, re.DOTALL | re.IGNORECASE)
        
        if xml_match:
            return xml_match.group(0)
        
        # Alternative search patterns
        xml_patterns = [
            r'<[^>]+>.*?</[^>]+>',  # Any XML-like content
            r'<\w+[^>]*>.*?</\w+>',  # Standard XML elements
        ]
        
        for pattern in xml_patterns:
            xml_match = re.search(pattern, content_str, re.DOTALL)
            if xml_match:
                xml_content = xml_match.group(0)
                # Check if it looks like valid XML
                if '<' in xml_content and '>' in xml_content:
                    return xml_content
        
        raise ValueError("No XML content found in PEM file")
        
    except Exception as e:
        raise RuntimeError(f"Failed to extract XML from PEM: {str(e)}")


def xml_to_dict(element: ET.Element, parent_key: str = "") -> Dict[str, Any]:
    """Convert XML element to dictionary with flattened structure."""
    result = {}
    
    # Add attributes
    for attr_name, attr_value in element.attrib.items():
        key = f"{parent_key}.@{attr_name}" if parent_key else f"@{attr_name}"
        result[key] = attr_value
    
    # Add text content
    if element.text and element.text.strip():
        text_key = f"{parent_key}.#text" if parent_key else "#text"
        result[text_key] = element.text.strip()
    
    # Group children by tag name
    children_by_tag = {}
    for child in element:
        tag = child.tag
        if tag not in children_by_tag:
            children_by_tag[tag] = []
        children_by_tag[tag].append(child)
    
    # Process children
    for tag, children in children_by_tag.items():
        if len(children) == 1:
            # Single child
            child_key = f"{parent_key}.{tag}" if parent_key else tag
            child_dict = xml_to_dict(children[0], child_key)
            result.update(child_dict)
        else:
            # Multiple children with same tag
            for i, child in enumerate(children):
                child_key = f"{parent_key}.{tag}[{i}]" if parent_key else f"{tag}[{i}]"
                child_dict = xml_to_dict(child, child_key)
                result.update(child_dict)
    
    return result


def dict_to_csv(data: Dict[str, Any], output_file: str) -> None:
    """Convert dictionary to CSV file."""
    if not data:
        raise ValueError("No data to convert to CSV")
    
    # Write CSV
    with open(output_file, 'w', newline='', encoding='utf-8') as csvfile:
        writer = csv.writer(csvfile)
        
        # Write header
        headers = list(data.keys())
        writer.writerow(headers)
        
        # Write data row
        values = [str(data.get(header, "")) for header in headers]
        writer.writerow(values)


def main():
    """Main function."""
    parser = argparse.ArgumentParser(description="Convert PEM files with XML content to CSV")
    parser.add_argument("--file", "-f", help="Path to PEM file to process")
    args = parser.parse_args()
    
    try:
        print("🚀 PEM to CSV Converter")
        print("========================\n")
        
        # Select PEM file
        pem_file_path = select_pem_file(args.file)
        print(f"\n📖 Reading PEM file: {pem_file_path}")
        
        # Read PEM file
        with open(pem_file_path, 'r', encoding='utf-8') as f:
            pem_content = f.read()
        print("✅ PEM file loaded successfully")
        
        # Extract XML
        print("🔍 Extracting XML content using ASN.1 parsing...")
        xml_content = extract_xml_from_pem(pem_content)
        print("✅ XML content extracted successfully")
        
        # Parse XML
        print("🔄 Parsing XML to dictionary structure...")
        try:
            root = ET.fromstring(xml_content)
            xml_dict = xml_to_dict(root)
        except ET.ParseError as e:
            print(f"⚠️  XML parsing error: {e}")
            # Try to clean up the XML
            xml_content = re.sub(r'[^\x20-\x7E\n\r\t]', '', xml_content)  # Remove non-printable chars
            root = ET.fromstring(xml_content)
            xml_dict = xml_to_dict(root)
        
        print("✅ XML parsed successfully")
        
        # Convert to CSV
        print("📊 Converting to CSV format...")
        output_file = Path(pem_file_path).stem + ".csv"
        dict_to_csv(xml_dict, output_file)
        print("✅ CSV conversion completed")
        
        print(f"\n🎉 Conversion completed successfully!")
        print(f"📁 Input:  {pem_file_path}")
        print(f"📁 Output: {output_file}")
        
        # Show preview
        print(f"\n📋 CSV Preview:")
        print("=" * 50)
        with open(output_file, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            for i, line in enumerate(lines[:5]):  # Show first 5 lines
                print(f"{i+1}: {line.strip()}")
            if len(lines) > 5:
                print(f"... and {len(lines) - 5} more lines")
        
    except KeyboardInterrupt:
        print("\n❌ Operation cancelled by user")
        sys.exit(1)
    except Exception as e:
        print(f"❌ Error: {str(e)}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: pem-to-csv.ts
#!/usr/bin/env -S deno run --allow-read --allow-write --allow-net

import { parse } from "https://deno.land/std@0.208.0/flags/mod.ts";
import { exists } from "https://deno.land/std@0.208.0/fs/mod.ts";
import { dirname, join, basename } from "https://deno.land/std@0.208.0/path/mod.ts";
import forge from "https://esm.sh/node-forge@1.3.1";

interface XMLNode {
  [key: string]: string | XMLNode | XMLNode[];
}

// Simple XML parser for converting to flat structure
function parseXMLToObject(xmlString: string): XMLNode {
  const parser = new DOMParser();
  const doc = parser.parseFromString(xmlString, "text/xml");
  
  function nodeToObject(node: Element): XMLNode {
    const obj: XMLNode = {};
    
    // Add attributes
    for (const attr of Array.from(node.attributes)) {
      obj[`@${attr.name}`] = attr.value;
    }
    
    // Add child nodes
    const children = Array.from(node.children);
    if (children.length === 0) {
      // Leaf node with text content
      const textContent = node.textContent?.trim();
      if (textContent) {
        obj["#text"] = textContent;
      }
    } else {
      for (const child of children) {
        const childName = child.tagName;
        const childObj = nodeToObject(child);
        
        if (obj[childName]) {
          // Multiple elements with same name - convert to array
          if (!Array.isArray(obj[childName])) {
            obj[childName] = [obj[childName] as XMLNode];
          }
          (obj[childName] as XMLNode[]).push(childObj);
        } else {
          obj[childName] = childObj;
        }
      }
    }
    
    return obj;
  }
  
  const rootElement = doc.documentElement;
  if (!rootElement) {
    throw new Error("Invalid XML: No root element found");
  }
  
  return { [rootElement.tagName]: nodeToObject(rootElement) };
}

// Flatten nested object to create CSV rows
function flattenObject(obj: any, prefix = ""): Record<string, string> {
  const flattened: Record<string, string> = {};
  
  for (const key in obj) {
    const value = obj[key];
    const newKey = prefix ? `${prefix}.${key}` : key;
    
    if (Array.isArray(value)) {
      value.forEach((item, index) => {
        if (typeof item === "object" && item !== null) {
          Object.assign(flattened, flattenObject(item, `${newKey}[${index}]`));
        } else {
          flattened[`${newKey}[${index}]`] = String(item);
        }
      });
    } else if (typeof value === "object" && value !== null) {
      Object.assign(flattened, flattenObject(value, newKey));
    } else {
      flattened[newKey] = String(value);
    }
  }
  
  return flattened;
}

// Convert object to CSV format
function objectToCSV(data: Record<string, string>[]): string {
  if (data.length === 0) return "";
  
  const headers = Array.from(new Set(data.flatMap(row => Object.keys(row))));
  const csvRows = [headers.join(",")];
  
  for (const row of data) {
    const values = headers.map(header => {
      const value = row[header] || "";
      // Escape quotes and wrap in quotes if contains comma, quote, or newline
      if (value.includes(",") || value.includes('"') || value.includes("\n")) {
        return `"${value.replace(/"/g, '""')}"`;
      }
      return value;
    });
    csvRows.push(values.join(","));
  }
  
  return csvRows.join("\n");
}

// Extract XML from PEM file using ASN.1
function extractXMLFromPEM(pemContent: string): string {
  try {
    // Handle different PEM formats
    let base64Content = pemContent;
    const pemHeaders = [
      "-----BEGIN CMS-----",
      "-----END CMS-----",
      "-----BEGIN PKCS7-----",
      "-----END PKCS7-----",
      "-----BEGIN CERTIFICATE-----",
      "-----END CERTIFICATE-----"
    ];
    
    // Remove PEM headers/footers
    for (const header of pemHeaders) {
      base64Content = base64Content.replace(header, "");
    }
    base64Content = base64Content.replace(/\s/g, "");

    const der = forge.util.decode64(base64Content);
    const asn1 = forge.asn1.fromDer(der);
    
    // Try to parse as PKCS#7/CMS first
    try {
      const p7 = forge.pkcs7.messageFromAsn1(asn1);
      const content = p7.content.toString();
      
      // Look for XML content
      const xmlMatch = content.match(/<\?xml[\s\S]*?<\/[^>]+>/i);
      if (xmlMatch) {
        return xmlMatch[0];
      }
    } catch (e) {
      console.warn("Failed to parse as PKCS#7, trying direct ASN.1 parsing");
    }
    
    // If PKCS#7 parsing fails, try to find XML in ASN.1 structure
    function searchASN1ForXML(obj: any): string | null {
      if (typeof obj === "string") {
        const xmlMatch = obj.match(/<\?xml[\s\S]*?<\/[^>]+>/i);
        if (xmlMatch) return xmlMatch[0];
      }
      
      if (obj && typeof obj === "object") {
        if (obj.value) {
          const result = searchASN1ForXML(obj.value);
          if (result) return result;
        }
        
        if (Array.isArray(obj)) {
          for (const item of obj) {
            const result = searchASN1ForXML(item);
            if (result) return result;
          }
        }
      }
      
      return null;
    }
    
    const xmlContent = searchASN1ForXML(asn1);
    if (xmlContent) {
      return xmlContent;
    }
    
    throw new Error("No XML content found in PEM file");
  } catch (error) {
    throw new Error(`Failed to extract XML from PEM: ${error.message}`);
  }
}

// Interactive file selection
async function selectPEMFile(): Promise<string> {
  const args = parse(Deno.args);
  
  if (args.file) {
    return args.file;
  }
  
  console.log("🔍 Looking for PEM files in current directory...");
  
  const pemFiles: string[] = [];
  for await (const entry of Deno.readDir(".")) {
    if (entry.isFile && entry.name.endsWith(".pem")) {
      pemFiles.push(entry.name);
    }
  }
  
  if (pemFiles.length === 0) {
    console.log("❌ No PEM files found in current directory");
    console.log("Usage: deno run --allow-read --allow-write pem-to-csv.ts --file <path-to-pem-file>");
    Deno.exit(1);
  }
  
  console.log("\n📁 Available PEM files:");
  pemFiles.forEach((file, index) => {
    console.log(`  ${index + 1}. ${file}`);
  });
  
  console.log("\n📝 Enter the number of the file you want to process (or press Enter for file 1):");
  
  const input = prompt("Selection: ") || "1";
  const selection = parseInt(input) - 1;
  
  if (selection < 0 || selection >= pemFiles.length) {
    console.log("❌ Invalid selection");
    Deno.exit(1);
  }
  
  return pemFiles[selection];
}

// Main function
async function main() {
  try {
    console.log("🚀 PEM to CSV Converter");
    console.log("========================\n");
    
    const pemFilePath = await selectPEMFile();
    
    console.log(`\n📖 Reading PEM file: ${pemFilePath}`);
    
    if (!await exists(pemFilePath)) {
      throw new Error(`File not found: ${pemFilePath}`);
    }
    
    const pemContent = await Deno.readTextFile(pemFilePath);
    console.log("✅ PEM file loaded successfully");
    
    console.log("🔍 Extracting XML content using ASN.1 parsing...");
    const xmlContent = extractXMLFromPEM(pemContent);
    console.log("✅ XML content extracted successfully");
    
    console.log("🔄 Parsing XML to object structure...");
    const xmlObject = parseXMLToObject(xmlContent);
    console.log("✅ XML parsed successfully");
    
    console.log("📊 Converting to CSV format...");
    const flattened = flattenObject(xmlObject);
    const csvData = objectToCSV([flattened]);
    console.log("✅ CSV conversion completed");
    
    // Generate output filename
    const baseName = basename(pemFilePath, ".pem");
    const outputPath = `${baseName}.csv`;
    
    console.log(`💾 Saving CSV to: ${outputPath}`);
    await Deno.writeTextFile(outputPath, csvData);
    
    console.log("\n🎉 Conversion completed successfully!");
    console.log(`📁 Input:  ${pemFilePath}`);
    console.log(`📁 Output: ${outputPath}`);
    
    // Show preview of CSV content
    const lines = csvData.split("\n");
    console.log("\n📋 CSV Preview (first 5 lines):");
    console.log("=" .repeat(50));
    lines.slice(0, 5).forEach(line => console.log(line));
    if (lines.length > 5) {
      console.log(`... and ${lines.length - 5} more lines`);
    }
    
  } catch (error) {
    console.error(`❌ Error: ${error.message}`);
    Deno.exit(1);
  }
}

// Run the script
if (import.meta.main) {
  await main();
}